src/pythie_serving/lightgbm_wrapper.py

```diff
@@ -36,7 +40,56 @@ def _create_model_specs(self, model_config: ModelConfig) -> ModelSpecs:
                 "feature_names": feature_names,
                 "nb_features": len(feature_names),
                 "samples_dtype": float,
+                "extra_specs": None
             }
 
     def _predict(self, model_specs: ModelSpecs, samples: NDArray) -> NDArray:
         return model_specs["model"].predict(samples, nthread=self.nthread)
+
+
+class LGBMCountServiceServicer(AbstractPythieServingPredictionServiceServicer):
+    model_file_extension = ".pickled"
+
+    def __init__(self, *, model_server_config: ModelServerConfig):
+        super().__init__(model_server_config=model_server_config)
+        self.nthread = int(os.environ.get("LGBM_NTHREAD", 0))
+
+    def _create_model_specs(self, model_config: ModelConfig) -> ModelSpecs:
+        with open(self._get_model_path(model_config), "rb") as opened_model:
+            model = pickle.load(opened_model)
+
+            if isinstance(model, Booster):
+                feature_names = model.feature_name()
+            else:
+                feature_names = model.feature_names
+
+        with open(self._get_metadata_path(model_config)) as f:
+            metadata = json.load(f)
+            leaf_index_to_sample_count = metadata["leaf_index_to_count"]
+
+        return {
+            "model": model,
+            "feature_names": feature_names,
+            "nb_features": len(feature_names),
+            "samples_dtype": float,
+            "extra_specs": leaf_index_to_sample_count
+        }
+
+    def _predict(self, model_specs: ModelSpecs, samples: NDArray) -> NDArray:
+        leaf_index_to_sample_count: Dict[str, int] = model_specs["extra_specs"]
+        nb_samples = len(samples)
+
+        leaf_indexes_in_trees: NDArray = model_specs["model"].predict(samples, nthread=self.nthread, pred_leaf=True)
+        nb_trees = leaf_indexes_in_trees.shape[1]
+        total_count = []
+        for sample_position in range(nb_samples): # for each sample get the number of samples
+            # used for its prediction
+            leafs_count = []
+            for tree_id in range(nb_trees):
+                leaf_index = f"{tree_id}-L{leaf_indexes_in_trees[sample_position, tree_id]}"
+                leaf_count = leaf_index_to_sample_count[leaf_index]
+                leafs_count.append(leaf_count)
+            total_count.append([np.mean(leafs_count)])
```

Comment:
`nb_trees` is number of trees in the whole model ? That's gonna be huge, might be really slow, no?

It think something like this might be slightly faster? (avoid appends, array conversion, etc)

```
total_count = np.array(
    np.mean(
        [leaf_index_to_sample_count[f"{tree_id}-L{leaf_indexes_in_trees[i, tree_id]}"] for tree_id in range(nb_trees)]
    )
    for i in range(nb_samples)
)
```

Reply:
Well I just avoid using list comprehension here to keep the code very easy to grasp. But ok I can use it.